merkle/sparse_merkle_tree.go

```diff
@@ -25,7 +27,191 @@ type SparseMerkleTreeReader struct {
 type SparseMerkleTreeWriter struct {
 	tx           storage.TreeTX
 	hasher       MapHasher
-	treeRevision uint64
+	treeRevision int64
+	tree         Subtree
+}
+
+type indexAndHash struct {
+	index []byte
+	hash  trillian.Hash
+}
+
+// rootHashOrError represents a (sub-)tree root hash, or an error which
+// prevented the calculation from completing.
+type rootHashOrError struct {
+	hash trillian.Hash
+	err  error
+}
+
+// Subtree is an interface which must be implemented by subtree workers.
+// Currently there's only a locally sharded go-routine based implementation,
+// the the idea is that an RPC based sharding implementation could be created
+// and dropped in.
+type Subtree interface {
+	// SetLeaf sets a single leaf hash for integration into a sparse merkle tree.
+	SetLeaf(index []byte, hash trillian.Hash) error
+
+	// CalculateRoot instructs the subtree worker to start calculating the root
+	// has of its tree.  It is an error to call SetLeaf() after calling this
+	// method.
+	CalculateRoot()
+
+	// RootHash returns the calculated root hash for this subtree, if the root
+	// hash has not yet been calculated, this method will block until it is.
+	RootHash() (trillian.Hash, error)
+}
+
+// getSubtreeFunc is essentially a factory method for getting child subtrees.
+type getSubtreeFunc func(prefix []byte) Subtree
+
+// subtreeWriter knows how to calculate and store nodes for a subtree.
+type subtreeWriter struct {
+	// prefix is the path to the root of this subtree in the full tree.
+	// i.e. all paths/indices under this tree share the same prefix.
+	prefix []byte
+
+	// subtreeDepth is the number of levels this subtree contains.
+	subtreeDepth int
+
+	// leafQueue is the work-queue containing leaves to be integrated into the
+	// subtree.
+	leafQueue chan func() (*indexAndHash, error)
+
+	// root is channel of size 1 from which the subtree root can be read once it
+	// has been calculated.
+	root chan rootHashOrError
+
+	// childMutex protexts access to children.
+	childMutex sync.RWMutex
+
+	// children is a map of child-subtrees by stringified prefix.
+	children map[string]Subtree
+
+	tx           storage.TreeTX
+	treeRevision int64
+
+	treeHasher TreeHasher
+
+	getSubtree getSubtreeFunc
+}
+
+// getOrCreateChildSubtree returns, or creates and returns, a subtree worker
+// for the specified childPrefix.
+func (s *subtreeWriter) getOrCreateChildSubtree(childPrefix []byte) Subtree {
+	childPrefixStr := string(childPrefix)
+	s.childMutex.Lock()
+	defer s.childMutex.Unlock()
+
+	subtree := s.children[childPrefixStr]
+	if subtree == nil {
+		subtree = s.getSubtree(childPrefix)
+		s.children[childPrefixStr] = subtree
+
+		// Since a new subtree worker is being created, then we'll add a future is
```

Comment:
yeah, that was gibberish - I've changed it a bit and hopefully it's a bit clearer.